From bc053bafdc1044830eee8bd88b464aa65d1965dc Mon Sep 17 00:00:00 2001
From: Marc Mao <xmao4@android.intel.com>
Date: Thu, 22 Apr 2021 20:56:02 +0800
Subject: [PATCH] Set GL error for glProgramBinary failed

For Android hwui, it will try to use persitent cache to save/restore
program binary. But if the driver updated, the previous cache will be
invalid since sha1 hash is changed with driver change. We need to return
GL error so that app can recompile the shader.

Signed-off-by: Marc Mao <xmao4@android.intel.com>
Change-Id: I98ce7ba0fa231f26a498473009cb358dec8335a1
---
 src/mesa/main/program_binary.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/mesa/main/program_binary.c b/src/mesa/main/program_binary.c
index 6aaa2c5a80e..4933d567397 100644
--- a/src/mesa/main/program_binary.c
+++ b/src/mesa/main/program_binary.c
@@ -283,6 +283,8 @@ _mesa_program_binary(struct gl_context *ctx, struct gl_shader_program *sh_prog,
 
    if (payload == NULL) {
       sh_prog->data->LinkStatus = LINKING_FAILURE;
+      _mesa_error(ctx, GL_INVALID_VALUE,
+               "glProgramBinary(invalid program binary)");
       return;
    }
 
@@ -300,6 +302,8 @@ _mesa_program_binary(struct gl_context *ctx, struct gl_shader_program *sh_prog,
 
    if (!read_program_payload(ctx, &blob, binary_format, sh_prog)) {
       sh_prog->data->LinkStatus = LINKING_FAILURE;
+      _mesa_error(ctx, GL_INVALID_VALUE,
+               "glProgramBinary(failed to read program binary)");
       return;
    }
 
-- 
2.33.1

